core/commands/dag/dag.go

```diff
@@ -241,3 +277,487 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+// It is not guaranteed that a root in a header is actually present in the same ( or any )
+// .car file. This is the case in version 1, and ideally in further versions too
+// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+// We will attempt a pin *only* at the end in case all car files were well formed
+//
+// The boolean value indicates whether we have encountered the root within the car file's
+type expectedRootsSeen map[cid.Cid]bool
```

Comment:
This type just obfuscates the underlying type and doesn't really gain us anything.

Reply:
It is a ternary:
- no idea
- seen mentioned as root
- seen mentioned as root and saw the corresponding payload of the block

Can you elaborate?

Reply:
I mean defining a custom `expectedRootsSeen` type when it's just a simple CID set just means I have to lookup the definition of `expectedRootsSeen` to understand what it is. The underlying type (`map[cid.Cid]bool`) makes sense.

Reply:
Fixed in https://github.com/ipfs/go-ipfs/pull/7038